fairing/functions/function_shim.py

```diff
@@ -1,10 +1,49 @@
 import argparse
 import cloudpickle
+import types
+from enum import Enum
+import logging
+
+logging.basicConfig(format='%(message)s')
+logging.getLogger().setLevel(logging.INFO)
+
+class ObjectType(Enum):
+    FUNCTION = 1
+    CLASS = 2
+    NOT_SUPPORTED = 3
+
+def get_execution_obj_type(obj):
+    # Check if a function is provided
+    if (isinstance(obj, types.FunctionType) or
+        isinstance(obj, types.BuiltinFunctionType) or
+        isinstance(obj, types.BuiltinMethodType)):
+        return ObjectType.FUNCTION
+
+    # Reject class methods provided directly
+    if isinstance(obj, types.MethodType):
+        return ObjectType.NOT_SUPPORTED
+
+    # Check if a class is provided (Python 2)
+    if (hasattr(types, 'ClassType') and isinstance(obj, types.ClassType)
+        and hasattr(obj, 'train')):
+        return ObjectType.CLASS
+
+    # Check if a class is provided (Python 3)
+    if isinstance(obj, type) and hasattr(obj, 'train'):
+        return ObjectType.CLASS
+
+    return ObjectType.NOT_SUPPORTED
 
 def call(serialized_fn_file):
     with open(serialized_fn_file, 'rb') as f:
-        fn = cloudpickle.load(f)
-        res = fn()
+        obj = cloudpickle.load(f)
+        obj_type = get_execution_obj_type(obj)
+        if obj_type == ObjectType.FUNCTION:
+            res = obj()
+        elif obj_type == ObjectType.CLASS:
```

Comment:
I'm confused why a user can't just pass in the class method themselves?

```
hs = HousingServe()
remote_train = fairing.config.fn(hs.train)
```

Reply:
If the train method is a class method then you can use it by passing "HousingServe.train" . If the train method is an instance method then we can't just serialize that as we need the object itself to invoke it in the remote environment, that's why we want the user to pass the class reference. The users can pass an instance of a class and we can call just train method but that behaviour can be done through just making train method class method or specifying it as an independent function.